File: usedesk_sdk/src/main/java/ru/usedesk/sdk/external/entity/chat/UsedeskActionListenerRx.java
package ru.usedesk.sdk.external.entity.chat;

import android.support.annotation.NonNull;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import io.reactivex.Observable;
import io.reactivex.subjects.BehaviorSubject;
import io.reactivex.subjects.Subject;
import ru.usedesk.sdk.external.entity.exceptions.UsedeskException;

public class UsedeskActionListenerRx implements UsedeskActionListener {

    private final Subject<EmptyItem> connectedSubject = BehaviorSubject.create();
    private final Subject<Message> messageSubject = BehaviorSubject.create();
    private final Subject<Message> newMessageSubject = BehaviorSubject.create();
    private final Subject<EmptyItem> offlineFormExpectedSubject = BehaviorSubject.create();
    private final Subject<EmptyItem> disconnectedSubject = BehaviorSubject.create();
    private final Subject<Integer> errorResIdSubject = BehaviorSubject.create();
    private final Subject<Exception> errorSubject = BehaviorSubject.create();
    private final Subject<UsedeskException> exceptionSubject = BehaviorSubject.create();

    private final Observable<List<Message>> messagesObservable;

    @Inject
    public UsedeskActionListenerRx() {
        messagesObservable = messageSubject.scan(new ArrayList<>(), (messages, message) -> {
            List<Message> newMessages = new ArrayList<>(messages);
            newMessages.add(message);
            return newMessages;
        });
    }

    private void onMessage(Message message) {
        if (message != null) {
            messageSubject.onNext(message);
        }
    }

    private void onNewMessage(Message message) {
        if (message != null) {
            newMessageSubject.onNext(message);
        }
    }

    public Observable<EmptyItem> getConnectedObservable() {
        return connectedSubject;
    }

    /**
     * Все сообщения
     */
    public Observable<Message> getMessageObservable() {
        return messageSubject;
    }

    /**
     * Только новые сообщения, генерируемые после подписки
     */
    public Observable<Message> getNewMessageObservable() {
        return newMessageSubject;
    }

    /**
     * Полный список сообщений, генерируется с каждым сообщением
     */
    public Observable<List<Message>> getMessagesObservable() {
        return messagesObservable;
    }

    public Observable<EmptyItem> getOfflineFormExpectedObservable() {
        return offlineFormExpectedSubject;
    }

    public Observable<EmptyItem> getDisconnectedSubject() {
        return disconnectedSubject;
    }

    public Observable<Integer> getErrorResIdSubject() {
        return errorResIdSubject;
    }

    public Observable<Exception> getErrorSubject() {
        return errorSubject;
    }

    @Override
    public void onConnected() {
        connectedSubject.onNext(EmptyItem.IGNORE_ME);
    }

    @Override
    public void onMessageReceived(Message message) {
        onMessage(message);
        onNewMessage(message);
    }

    @Override
    public void onMessagesReceived(List<Message> messages) {
        if (messages != null) {
            for (Message message : messages) {
                onMessage(message);
            }
        }
    }

    @Override
    public void onServiceMessageReceived(Message message) {
        onMessage(message);
    }

    @Override
    public void onOfflineFormExpected() {
        connectedSubject.onNext(EmptyItem.IGNORE_ME);
    }

    @Override
    public void onDisconnected() {
        connectedSubject.onNext(EmptyItem.IGNORE_ME);
    }

    @Override
    @Deprecated
    public void onError(int errorResId) {
        errorResIdSubject.onNext(errorResId);
    }

    @Override
    @Deprecated
    public void onError(Exception e) {
        if (e != null) {
            errorSubject.onError(e);
        }
    }

    @Override
    public void onException(@NonNull UsedeskException usedeskException) {
        exceptionSubject.onNext(usedeskException);
    }

    public Subject<UsedeskException> getExceptionSubject() {
        return exceptionSubject;
    }

    public enum EmptyItem {
        IGNORE_ME
    }
}

File: chat-sdk/src/main/java/ru/usedesk/chat_sdk/entity/UsedeskActionListenerRx.kt
package ru.usedesk.chat_sdk.entity

import io.reactivex.Observable
import io.reactivex.subjects.BehaviorSubject

class UsedeskActionListenerRx : IUsedeskActionListener {

    private val connectedSubject = BehaviorSubject.create<UsedeskEvent<Any?>>()
    val connectedObservable: Observable<UsedeskEvent<Any?>> = connectedSubject

    private val disconnectedSubject = BehaviorSubject.create<UsedeskEvent<Any?>>()
    val disconnectedObservable: Observable<UsedeskEvent<Any?>> = disconnectedSubject

    private val connectedStateSubject = BehaviorSubject.createDefault(false)
    val connectedStateObservable: Observable<Boolean> = connectedStateSubject

    private val messageSubject = BehaviorSubject.create<UsedeskMessage>()
    val messageObservable: Observable<UsedeskMessage> = messageSubject//Каждое сообщение по отдельности

    private val newMessageSubject = BehaviorSubject.create<UsedeskMessage>()
    val newMessageObservable: Observable<UsedeskMessage> = newMessageSubject//Только новые сообщения, генерируемые после подписки

    private val messagesSubject = BehaviorSubject.create<List<UsedeskMessage>>()
    val messagesObservable: Observable<List<UsedeskMessage>> = messagesSubject//Список всех сообщений (обновляется с каждым новым сообщением)

    private val messageUpdateSubject = BehaviorSubject.create<UsedeskMessage>()
    val messageUpdateObservable: Observable<UsedeskMessage> = messageUpdateSubject//Сообщение, изменившее своё состояние

    private val offlineFormExpectedSubject = BehaviorSubject.create<UsedeskChatConfiguration>()
    val offlineFormExpectedObservable: Observable<UsedeskChatConfiguration> = offlineFormExpectedSubject

    private val feedbackSubject = BehaviorSubject.create<UsedeskEvent<Any?>>()
    val feedbackObservable: Observable<UsedeskEvent<Any?>> = feedbackSubject

    private val exceptionSubject = BehaviorSubject.create<Exception>()
    val exceptionObservable: Observable<Exception> = exceptionSubject

    private var lastMessages = listOf<UsedeskMessage>()

    private fun onNewMessages(newMessages: List<UsedeskMessage>) {
        postMessages(lastMessages + newMessages)
    }

    private fun postMessages(messages: List<UsedeskMessage>) {
        lastMessages = messages
        messagesSubject.onNext(messages)
    }

    override fun onConnected() {
        connectedSubject.onNext(UsedeskSingleLifeEvent(null))
        connectedStateSubject.onNext(true)
    }

    override fun onMessageReceived(message: UsedeskMessage) {
        messageSubject.onNext(message)
        newMessageSubject.onNext(message)
        onNewMessages(listOf(message))
    }

    override fun onMessagesReceived(messages: List<UsedeskMessage>) {
        messages.forEach {
            messageSubject.onNext(it)
        }
        postMessages(messages)
    }

    override fun onMessageUpdated(message: UsedeskMessage) {
        messageUpdateSubject.onNext(message)
    }

    override fun onFeedbackReceived() {
        feedbackSubject.onNext(UsedeskSingleLifeEvent(null))
    }

    override fun onOfflineFormExpected(chatConfiguration: UsedeskChatConfiguration) {
        offlineFormExpectedSubject.onNext(chatConfiguration)
    }

    override fun onDisconnected() {
        connectedSubject.onNext(UsedeskSingleLifeEvent(null))
        connectedStateSubject.onNext(false)
    }

    override fun onException(usedeskException: Exception) {
        exceptionSubject.onNext(usedeskException)
    }
}
